### ISSUE TITLE
QCStatement test failures w/ latest ZCrypto

### ISSUE THREAD

**Issue by cpu** [zmap/zlint#433]
The [`testdata/QcStmtEtsiTaggedValueCert20.pem`](https://github.com/zmap/zlint/blob/master/v2/testdata/QcStmtEtsiTaggedValueCert20.pem) test case used by [`TestEtsiTypeAsQcStmt`](https://github.com/zmap/zlint/blob/fd40f579253ea1ebfb18a585ab5cd8e7dcde61aa/v2/lints/etsi/lint_qcstatem_etsi_type_as_statem_test.go#L24-L41) panics during `test.ReadTestCert` when using ZCrypto w/ https://github.com/zmap/zcrypto/commit/13cf40b46ed3697eba66054aca90c73bbbc991d4. Here's [test output](https://gist.github.com/cpu/6cf8ac87958059274b5b4f2ccfa5d3d1) from a local repro. I also [pushed the branch](https://github.com/cpu/zlint/tree/cpu-bump-zcrypto).

The underlying error returned by ZCrypto's `x509.ParseCertificate` is an `encoding/asn1.StructuralError` generated trying to [`asn1.Unmarshal` the `oidExtQCStatements` extension value](https://github.com/zmap/zcrypto/blob/16679db567ff92527e6982f60286c63666054d0f/x509/x509.go#L1945-L1949) from the test cert into a `QCStatementsASN{}`: 

```
asn1: structure error: tags don't match (6 vs {class:2 tag:0 length:6 isCompound:false}) {optional:false explicit:false application:false private:false defaultValue:<nil> tag:<nil> stringType:0 timeType:0 set:false omitEmpty:false} ObjectIdentifier @2
```

Edit: There are a handful of other invalid test cases that fail w/ ZCrypto update: https://github.com/zmap/zlint/issues/433#issuecomment-628715654

~@mtgag This test case was introduced by you in https://github.com/zmap/zlint/commit/0f862af0a6db1b6bf2f7fffef5cb160583b93537 but there isn't a comment indicating what is being tested. Can you coordinate with @dadrian to figure out what needs to happen here? E.g. is this a test case that should be removed from ZLint or is it a bug with the ZCrypto QCStatement support? Thanks!~

**Comment by sleevi** [comment]
@cpu Thanks for opening. I can confirm that the cert is invalid.

```
      QCStatements ::= SEQUENCE OF QCStatement
      QCStatement ::= SEQUENCE {
          statementId   QC-STATEMENT.&Id({SupportedStatements}),
          statementInfo QC-STATEMENT.&Type
          ({SupportedStatements}{@statementId}) OPTIONAL }
```

The QC-STATEMENT class is defined at
```
   QC-STATEMENT ::= CLASS {
       &id   OBJECT IDENTIFIER UNIQUE,
       &Type OPTIONAL }
       WITH SYNTAX {
       [SYNTAX &Type] IDENTIFIED BY &id }
```

**Comment by cpu** [comment]
@sleevi Thanks!

There isn't really a way for ZLint to lint certificates that ZCrypto won't parse so I'm inclined to remove this test case and calling it done. @dadrian @sleevi Do you folks concur? Alternatively I think filing an issue on ZCrypto to determine next steps would be the best path forward.

**Comment by zakird** [comment]
I agree, we should remove this. It doesn't test code in ZLint as it stands.

**Comment by sleevi** [comment]
Agreed.

**Comment by cpu** [comment]
It looks like I made my issue too specific. There are other panics hidden behind this first one. Here's a few other testcase/testfile pairings that panic w/ ZCrypto's QCStatement parsing:

* `TestEtsiQcCompliance`, `testdata/QcStmtEtsiWrongEncodingCert01.pem`
* `TestEtsiQcType`, `testdata/QcStmtEtsiMissingSeqForQcTypesCert18.pem`
* `TestEtsiQcType`, `testdata/QcStmtEtsiMissingOidCert09.pem`

I'll whack these as well.

**Comment by mtgag** [comment]
Hello,

went back to the code that produced the certificates and the comment for this certificate is "cotaining something tagged". The intention was to deliberately build a certificate with a wrong encoding, in this case one tagged type which does appear in this extension. This use case is somehow constructed and could be removed.

**Comment by sleevi** [comment]
Yeah, I'd be supportive of moving these to ZCrypto, which helps test that they enforce the ASN.1 requirements correctly and reject "obviously junk" certs. When ZCrypto didn't parse these, it makes sense to make sure ZLint would yell at these obviously bad certs.

**Comment by mtgag** [comment]
> 
> 
> It looks like I made my issue too specific. There are other panics hidden behind this first one. Here's a few other testcase/testfile pairings that panic w/ ZCrypto's QCStatement parsing:
> 
>     * `TestEtsiQcCompliance`, `testdata/QcStmtEtsiWrongEncodingCert01.pem`
> 
>     * `TestEtsiQcType`, `testdata/QcStmtEtsiMissingSeqForQcTypesCert18.pem`
> 
>     * `TestEtsiQcType`, `testdata/QcStmtEtsiMissingOidCert09.pem`
> 
> 
> I'll whack these as well.

Yes, I expect more of them, also for the upcoming pull requests regarding ETSI, we create "broken" certificates for testing purposes. I guess this is something we should avoid.

**Comment by cpu** [comment]
> Yeah, I'd be supportive of moving these to ZCrypto, which helps test that they enforce the ASN.1 requirements correctly and reject "obviously junk" certs.

SGTM. I can take care of opening a PR for that as long as folks are happy with fairly surface level tests. I'm very keen on not spending too much of my free time actually learning about QCStatements :-)

**Comment by zakird** [comment]
That sounds good to me. That way we at least don't lose the code.

**Comment by cpu** [comment]
I'm working on getting a PR ready to update the ZCrypto version. Should be done shortly.

I removed the test cases flagged in this issue and unit tests are passing. Working on the integration test data now. Here's [an example of a certificate](https://censys.io/certificates/4712f1b2a94994b55626ecba2104bbf23d39c05e7a2751e5af8a923bac23fd8f) from the integration test data that no longer parses in case that's interesting to anyone.

One other note: for the time being I'm just getting the codebase working with the updated ZCrypto. I'm not going to refactor the existing QCStatement lints to use the new upstream parsing. I filed https://github.com/zmap/zlint/issues/434 for that part and would appreciate if someone could pick that up.
